### PR TITLE
Implement batch history replication API

### PR DESCRIPTION
**Pull request by wxing1292**
**What changed?**
* Implement batch history replication API

**Why?**
New API

**How did you test it?**
Existing UT, since logic is not really changed

**Potential risks**
N/A

**Is hotfix candidate?**
N/A

### CODE REVIEW

**Comment by wxing1292** [comment]
also refer #4110

File: service/history/workflow/mutable_state_rebuilder.go [hunk 1]
```diff
@@ -90,6 +90,36 @@ func NewMutableStateRebuilder(
...
+	history [][]*historypb.HistoryEvent,
+	newRunHistory []*historypb.HistoryEvent,
+) (MutableState, error) {
+	for i := 0; i < len(history)-1; i++ {
+		_, err := b.applyEvents(ctx, namespaceID, requestID, execution, history[i], nil)
+		if err != nil {
+			return nil, err
+		}
+	}
+	newMutableState, err := b.applyEvents(ctx, namespaceID, requestID, execution, history[len(history)-1], newRunHistory)
```
Comment (yux0) [inline service/history/workflow/mutable_state_rebuilder.go:106]: Sanity check len(history) - 1?

**Review by yux0 — COMMENTED** [review]
LGTM

File: service/history/workflow/mutable_state_rebuilder.go (hunk 1 above)
Comment (wxing1292) [reply service/history/workflow/mutable_state_rebuilder.go:106]: this is correct

the above for loop loops over all event batch except the last one
then use the last event batch and new run event, to generate the new ms iff continue as new

Comment (hehaifengcn) [reply service/history/workflow/mutable_state_rebuilder.go:108]: I think what YX asking was if `len(history) > 0` is always true. I can't infer that based on the code.

File: service/history/workflow/mutable_state_rebuilder.go [hunk 2]
```diff
@@ -90,6 +92,36 @@ func NewMutableStateRebuilder(
...
+	if err != nil {
+		return nil, err
+	}
+	// must generate the activity timer / user timer at the very end
```
Comment (hehaifengcn) [inline service/history/workflow/mutable_state_rebuilder.go:112]: can you entail why has to be very end?

File: service/history/workflow/mutable_state_rebuilder_test.go [hunk 3]
```diff
@@ -157,8 +157,8 @@ func (s *stateBuilderSuite) mockUpdateVersion(events ...*historypb.HistoryEvent)
 	s.mockMutableState.EXPECT().SetHistoryBuilder(NewImmutableHistoryBuilder(events))
 }
 
-func (s *stateBuilderSuite) toHistory(events ...*historypb.HistoryEvent) []*historypb.HistoryEvent {
-	return events
+func (s *stateBuilderSuite) toHistory(eventss ...*historypb.HistoryEvent) [][]*historypb.HistoryEvent {
```
Comment (hehaifengcn) [inline service/history/workflow/mutable_state_rebuilder_test.go:160]: I felt the mixed use of `...` and `[]` weird and not really consistent. is it some sort of convention on when to use `...` and when to use `[]`?

File: service/history/workflow/transaction_impl.go [hunk 4]
```diff
@@ -251,18 +251,31 @@ func (t *TransactionImpl) SetWorkflowExecution(
...
+			if err != nil {
+				return 0, err
+			}
+			totalSize += size
+		} else {
+			size, err := persistNonFirstWorkflowEvents(ctx, shard, workflowEvents)
+			if err != nil {
+				return 0, err
+			}
+			totalSize += size
```
Comment (hehaifengcn) [inline service/history/workflow/transaction_impl.go:275]: nit: you could move 
```
			if err != nil {
				return 0, err
			}
			totalSize += size
``` out of the if..else block and reduce duplicated code.

File: service/history/ndc/history_replicator.go [hunk 5]
```diff
@@ -96,11 +98,28 @@ type (
...
+		// NOTE:
+		//  1. all history events should have the same version
+		//  2. all history events should share the same version history
+		ApplyEventBlobs(
+			ctx context.Context,
+			workflowKey definition.WorkflowKey,
+			baseExecutionInfo *workflowpb.BaseExecutionInfo,
+			versionHistoryItems []*historyspb.VersionHistoryItem,
+			events [][]*historypb.HistoryEvent,
+			newEvents []*historypb.HistoryEvent,
```
Comment (hehaifengcn) [inline service/history/ndc/history_replicator.go:121]: maybe add a comment on what the difference between `events` and `newEvents`?

File: service/history/ndc/history_replicator.go [hunk 6]
```diff
@@ -666,9 +707,20 @@ func (r *HistoryReplicatorImpl) applyNonStartEventsToNonCurrentBranchWithoutCont
 		return err
 	}
 
+	eventsSlice := make([]*persistence.WorkflowEvents, len(task.getEvents()))
+	for i, events := range task.getEvents() {
+		eventsSlice[i] = &persistence.WorkflowEvents{
```
Comment (hehaifengcn) [inline service/history/ndc/history_replicator.go:712]: curious why not just do `eventSlice = append(eventSlice, ..)`

File: service/history/ndc/replication_task.go [hunk 7]
```diff
@@ -103,10 +103,9 @@ var (
 	ErrLastEventIsNotContinueAsNew = serviceerror.NewInvalidArgument("last event is not continue as new")
 )
 
-func newReplicationTask(
+func newReplicationTaskFromRequest(
 	clusterMetadata cluster.Metadata,
 	historySerializer serialization.Serializer,
-	taskStartTime time.Time,
```
Comment (hehaifengcn) [inline service/history/ndc/replication_task.go:109]: can you help me understand why `taskStartTime` is no longer needed?

File: service/history/workflow/mutable_state_rebuilder.go (hunk 1 above)
Comment (wxing1292) [reply service/history/workflow/mutable_state_rebuilder.go:108]: i see

input validation should be done at the beginning of the "API", e.g. at frontend service, at network handler.
this piece of logic is internal, so should trust basic input validation is already done.

File: service/history/workflow/context.go [hunk 8]
```diff
@@ -88,7 +88,7 @@ type (
 
 		PersistWorkflowEvents(
 			ctx context.Context,
-			workflowEvents *persistence.WorkflowEvents,
+			workflowEventsSlice ...*persistence.WorkflowEvents,
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply service/history/workflow/context.go:91]: easy to use, see https://gobyexample.com/variadic-functions

File: service/history/workflow/history_builder.go [hunk 9]
```diff
@@ -127,9 +127,10 @@ func NewMutableHistoryBuilder(
 }
 
 func NewImmutableHistoryBuilder(
-	history []*historypb.HistoryEvent,
+	histories ...[]*historypb.HistoryEvent,
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply service/history/workflow/history_builder.go:130]: `...[]*historypb.HistoryEvent,` is still an array of array, but easier for caller to make function call
so instead of 
```go
var events []*historypb.HistoryEvent
NewImmutableHistoryBuilder([][]*historypb.HistoryEvent{events})
```	

we can do
```go
```go
var events []*historypb.HistoryEvent
NewImmutableHistoryBuilder(events)
```

File: service/history/workflow/mutable_state_rebuilder.go [hunk 10]
```diff
@@ -90,6 +92,36 @@ func NewMutableStateRebuilder(
...
 
 func (b *MutableStateRebuilderImpl) ApplyEvents(
+	ctx context.Context,
+	namespaceID namespace.ID,
+	requestID string,
+	execution commonpb.WorkflowExecution,
+	history [][]*historypb.HistoryEvent,
+	newRunHistory []*historypb.HistoryEvent,
+) (MutableState, error) {
+	for i := 0; i < len(history)-1; i++ {
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply service/history/workflow/mutable_state_rebuilder.go:102]: the last events may create a new mutable state, see the additional invocation after the for loop

File: service/history/workflow/mutable_state_rebuilder.go (hunk 2 above)
Comment (wxing1292) [reply service/history/workflow/mutable_state_rebuilder.go:112]: this is how activity & user timer works, see https://github.com/temporalio/temporal/blob/master/service/history/workflow/timer_sequence.go

File: service/history/workflow/mutable_state_rebuilder_test.go (hunk 3 above)
Comment (wxing1292) [reply service/history/workflow/mutable_state_rebuilder_test.go:160]: 1. this is test function
2. ... vs [] actually depends on the business logic context: here, the logic need to apply []*historypb.HistoryEvent as a transaction, and we are trying to combine multiple transactions into save IO

File: service/history/workflow/transaction_impl.go (hunk 4 above)
Comment (wxing1292) [reply service/history/workflow/transaction_impl.go:275]: my main concern is not the duplicated code, but the `:=` from go, sometimes the var can be shadow-ed, causing unwanted behavior, e.g. https://github.com/temporalio/temporal/pull/1187

err defined here: https://github.com/temporalio/temporal/pull/1187/files#diff-72a05e1636a2bcea965072880e2c16672336b99edab8cf54c0add86e5877c09eL109

err shadow-ed here: https://github.com/temporalio/temporal/pull/1187/files#diff-72a05e1636a2bcea965072880e2c16672336b99edab8cf54c0add86e5877c09eL112

shadow-ed err assignment here:https://github.com/temporalio/temporal/pull/1187/files#diff-72a05e1636a2bcea965072880e2c16672336b99edab8cf54c0add86e5877c09eL127

err check here: https://github.com/temporalio/temporal/pull/1187/files#diff-72a05e1636a2bcea965072880e2c16672336b99edab8cf54c0add86e5877c09eL135

basically err is defined outside of the if block, got shadowed & assigned within the if block; later when the error check is done from outside the if block, the err is *not* the updated err from within the if block.

File: service/history/ndc/history_replicator.go (hunk 6 above)
Comment (wxing1292) [reply service/history/ndc/history_replicator.go:712]: type conversion?

File: service/history/ndc/replication_task.go (hunk 7 above)
Comment (wxing1292) [reply service/history/ndc/replication_task.go:109]: the variable is no longer used by downstream APIs

**Review by meiliang86 — APPROVED** [review]
